sdk/python/feast/feature_store.py

```diff
@@ -706,6 +709,11 @@ def get_historical_features(
             full_feature_names: A boolean that provides the option to add the feature view prefixes to the feature names,
                 changing them from the format "feature" to "feature_view__feature" (e.g., "daily_transactions" changes to
                 "customer_fv__daily_transactions"). By default, this value is set to False.
+            save_as: An optional configuration that enables persisting retrieved dataset in offline store.
```

Comment:
Isn't this API overlapping with [RetrievalJob.to_bigquery()](https://github.com/feast-dev/feast/blob/97dd41e0c0e1a9e8050f3931eb610f59789edeed/sdk/python/feast/infra/offline_stores/bigquery.py#L255) etc? I think it's a bit confusing to me to have two places to tell Feast to save a dataset.

Reply:
Yes, you can call `job.to_bigquery()` or `job.to_redshift()` and this is what saved dataset implementation does under the hood. The difference is that you will need to add additional DataSource in order to read those saved tables back to Feast. And this is what this PR does automatically via `store.get_saved_dataset`

Reply:
it also is slightly unclear to me what the exact contract of RetrievalJob is. RetrievalJob otherwise is a deferred fetch that can be triggered on e.g. to_df() or to_arrow(). 

The save_as parameter seems to be ambiguous as to when it persists the data. (the implementation is it happens immediately). If we want the timing to be consistent, then it becomes very similar to the to_bigquery way of doing things.

Reply:
1. How it works: you're right that the original query will be triggered eagerly and the dataset will be persisted right away after calling `get_historical_features`. The caller will receive a substitution of RetrievalJob that will wrap a much simpler query: `select * from persisted_table` and process (contract) of materializing this query (job) remains the same (will be triggered on `.to_df()` or `.to_arrow()`).
2. Why not postpone persisting until the user will call `to_df()`: current implementation doesn't cache query results, it means that if I'll call `.to_df()` 3 times - 3 jobs to BigQuery or Redshift would be created. In case we'll put persisting on that stage - we'll have to overwrite the destination table over and over, which I don't think is good contract. Ideally, we should write once and fail if the table already exists and not risk deleting data created by the user and not Feast.